### PR TITLE
[W10H01] Added TrialOfDreams tests

### PR DESCRIPTION
**Pull request by mbmueller**

### CODE REVIEW

File: w10h01/test/pgdp/trials/TrialOfTheDreamsTest.java [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+        for (int i = 0; i < combination.length; i++) {
+            assertEquals(combination[i], result[i]);
+        }
+    }
+
+    @Test
+    @DisplayName("Combination of Length 3")
+    void TestLengthThree() {
+        byte[] combination = {25, 8, 30};
+        Function<byte[], Boolean> lock = new Function<byte[], Boolean>() {
```
Comment (Anatoly03) [inline w10h01/test/pgdp/trials/TrialOfTheDreamsTest.java:61]: Location 3

**Review by Anatoly03 — CHANGES_REQUESTED** [review]
- [How can I write an anonymous function in Java? - Stack Overflow] (https://stackoverflow.com/questions/2755445/how-can-i-write-an-anonymous-function-in-java)
- [Lambda Expressions Java](https://docs.oracle.com/javase/tutorial/java/javaOO/lambdaexpressions.html)

File: w10h01/test/pgdp/trials/TrialOfTheDreamsTest.java [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+import org.junit.jupiter.api.Test;
+
+import java.util.function.Function;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+public class TrialOfTheDreamsTest {
+    private Function<byte[], Boolean> getLock() {
+        return (byte[] bytes) -> {
+            if (bytes.length != combination.length) {return false;}
```
Comment (Anatoly03) [inline w10h01/test/pgdp/trials/TrialOfTheDreamsTest.java:13]: `combination` does not exist in current context

File: w10h01/test/pgdp/trials/TrialOfTheDreamsTest.java [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+            }
+            return true; 
+        }
+    }
+
+    @Test
+    @DisplayName("Combination of Length 1")
+    void TestLengthOne() {
+        byte[] combination = {Byte.MAX_VALUE};
+        Function<byte[], Boolean> lock = getLock();
```
Comment (Anatoly03) [inline w10h01/test/pgdp/trials/TrialOfTheDreamsTest.java:25]:
```suggestion
        Function<byte[], Boolean> lock = getLock(combination);
```
You can provide combination with an argument.

File: w10h01/test/pgdp/trials/TrialOfTheDreamsTest.java (hunk 2 above)
Comment (mbmueller) [reply w10h01/test/pgdp/trials/TrialOfTheDreamsTest.java:13]: damn it, you are rigth....

**Review by Anatoly03 — CHANGES_REQUESTED** [review]
Needs fix

**Comment by mbmueller** [comment]
> Needs fix

Working on it

File: w10h01/test/pgdp/trials/TrialOfTheDreamsTest.java [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+    void testEmpty() {
+        byte[] combination = new byte[0];
+        Function<byte[], Boolean> lock = getLock(combination);
+
+        byte[] result = TrialOfTheDreams.lockPick(lock);
+
+        for (int i = 0; i < combination.length; i++) {
+            assertEquals(combination[i], result[i]);
+        }
+    }
```
Comment (Anatoly03) [inline w10h01/test/pgdp/trials/TrialOfTheDreamsTest.java:84]:
```suggestion
    @ParameterizedTest
    @MethodSource
    public void combinationsTest(combination) {
        Function<byte[], Boolean> lock = getLock(combination);

        byte[] result = TrialOfTheDreams.lockPick(lock);

        for (int i = 0; i < combination.length; i++) {
            assertEquals(combination[i], result[i]);
        }
    }
    
    private static Stream<Arguments> combinationsTest {
        return Stream.of(
                arguments(
                    new byte[] {Byte.MAX_VALUE},
                    new byte[] {-128, 127},
                    new byte[] {25, 8, 30},
                    new byte[] {25, -8, 30},
                    new byte[0]
                ));
    }
```
You should also avoid code repetition by using tests from source. This suggestions comes with imports.

File: w10h01/test/pgdp/trials/TrialOfTheDreamsTest.java [hunk 5]
```diff
@@ -0,0 +1,85 @@
+package pgdp.trials;
+
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.Test;
+
```
Comment (Anatoly03) [inline w10h01/test/pgdp/trials/TrialOfTheDreamsTest.java:5]:
```suggestion
import java.util.stream.Stream;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.Arguments;
import static org.junit.jupiter.params.provider.Arguments.arguments;
import org.junit.jupiter.params.provider.MethodSource;
```

**Review by Anatoly03 — CHANGES_REQUESTED** [review]
You can also use `MethodSource` annotation.

**Comment by mbmueller** [comment]
Working on fix

**Comment by mbmueller** [comment]
I removed now the repetition but for some reason the parameterizedTest did not work, so I used a stream of byte[] and iterated through it using forEach

**Comment by mbmueller** [comment]
Please approve the change, as the Test is now working and can therefore help others figuring out the solution. Cosmetic optimizations can be made later IMHO.

File: w10h01/test/pgdp/trials/TrialOfTheDreamsTest.java [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+            if (bytes.length != combination.length) {return false;}
+            for (int i = 0; i < bytes.length; i++) {
+                if (bytes[i] != combination[i]) {return false;}
+            }
+            return true;
+        };
+    }
+
+    @Test
+    public void combinationsTest() {
```
Comment (Anatoly03) [inline w10h01/test/pgdp/trials/TrialOfTheDreamsTest.java:28]:
```suggestion
    @MethodSource
    @ParameterizedTest
    public void combinationsTest(byte[] combination) {
```

File: w10h01/test/pgdp/trials/TrialOfTheDreamsTest.java [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+        combinationsTestValues().forEach(combination -> {
+            Function<byte[], Boolean> lock = getLock(combination);
+            String msg = "For combination: " + Arrays.toString(combination);
+
+            byte[] result = TrialOfTheDreams.lockPick(lock);
+
+            for (int i = 0; i < combination.length; i++) {
+                assertEquals(combination[i], result[i], msg);
+            }
+        });
```
Comment (Anatoly03) [inline w10h01/test/pgdp/trials/TrialOfTheDreamsTest.java:39]:
```suggestion
        Function<byte[], Boolean> lock = getLock(combination);
        String msg = "For combination: " + Arrays.toString(combination);

        byte[] result = TrialOfTheDreams.lockPick(lock);

        for (int i = 0; i < combination.length; i++) {
            assertEquals(combination[i], result[i], msg);
        }
```

File: w10h01/test/pgdp/trials/TrialOfTheDreamsTest.java (hunk 6 above)
Comment (Anatoly03) [reply w10h01/test/pgdp/trials/TrialOfTheDreamsTest.java:28]: We declare it with `@ParameterizedTest` Annotation. This allows us to call tests through a function. `@MethodSource` declares that input will be a static function of same name but empty signature.

File: w10h01/test/pgdp/trials/TrialOfTheDreamsTest.java (hunk 7 above)
Comment (Anatoly03) [reply w10h01/test/pgdp/trials/TrialOfTheDreamsTest.java:39]: Hence we don't need `forEach`

File: w10h01/test/pgdp/trials/TrialOfTheDreamsTest.java [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+
+    private static Stream<byte[]> combinationsTestValues () {
+        return Stream.of(
+                    new byte[] {Byte.MAX_VALUE},
+                    new byte[] {-128, 127},
+                    new byte[] {25, 8, 30},
+                    new byte[] {25, -8, 30},
+                    new byte[0]
+                );
+    }
```
Comment (Anatoly03) [inline w10h01/test/pgdp/trials/TrialOfTheDreamsTest.java:50]:
```suggestion
    private static Stream<Arguments> combinationsTest() {
        return Stream.of(
                arguments(new byte[] { Byte.MAX_VALUE }),
                arguments(new byte[] { -128, 127 }),
                arguments(new byte[] { 25, 8, 30 }),
                arguments(new byte[] { 25, -8, 30 }),
                arguments(new byte[0]));
    }
```
The last thing we do is change the the method source to arguments type.

**Review by Anatoly03 — CHANGES_REQUESTED** [review]
> Please approve the change, as the Test is now working and can therefore help others figuring out the solution. Cosmetic optimizations can be made later IMHO.

@mbmueller I've went through your code and add the parameterized tests, only three parts need a change, I've added them! After these additions I'll approve gladly! Thank you for your patience!

**Review by Anatoly03 — APPROVED** [review]
LGTM!

**Comment by felixhauptmann** [comment]
<3 @Anatoly03

LGTM

**Pull request merged**